### PR TITLE
patch to always run exception handlers

### PR DESCRIPTION
**Pull request by Igorshp**
Patch for issue https://github.com/chef/chef/issues/3180

Actual problem:
missing recipe present in run_list - exception handlers work as expected
missing role present in run_list - exception handlers do not fire.

Exception handlers require @run_status to be set.
Moving the declaration to initialisation block ensures it's always available.

'node' is nil at that stage and is set again here https://github.com/chef/chef/blob/master/lib/chef/client.rb#L244

big thanks to @jonlives, Coderanger and @thommay your help

### CODE REVIEW

**Comment by thommay** [comment]
As I said on IRC, I think the patch is good but I'd love to see a test.

**Comment by Igorshp** [comment]
@thommay added a test for exact problem that started the PR. To ensure that exception handlers get run if we fail any time inside 'begin' block of run function.

**Comment by thommay** [comment]
:+1: this looks good to me

File: spec/unit/client_spec.rb [hunk 1]
```diff
@@ -819,4 +819,24 @@ def stub_for_audit
     end
 
   end
+
+  describe "should always attempt to run handlers" do
+    let(:e) { NoMethodError }
```
Comment (ranjib) [inline spec/unit/client_spec.rb]: nitpick, remove the `should`

**Comment by ranjib** [comment]
:+1:

**Comment by lamont-granquist** [comment]
Are you missing dropping the second declaration out of `#build_node`?

**Comment by Igorshp** [comment]
@lamont-granquist no, build node mutates @node and passes it to @run_status.
the declaration inside #build_node is correct.

This patch is simply to allow exception handlers to fire before that happens, even if full node data is not available. (from what i've seen, it doesn't impact messaging in any way)

**Comment by lamont-granquist** [comment]
So build_node actually creates the node object, and then build_node is responsible for wiring it up to the RunStatus.  The concern that I've got is that to do that, you are throwing away the old RunStatus object and creating a new one to wire it up.  Right now there's no state being saved in the first RunStatus object, but it seems like its asking for trouble to be doing that.  I'd probably be better to introduce a setter on the RunStatus to send it the node object in #build_node rather than toss away the first RunStatus object and create a new one.

**Comment by Igorshp** [comment]
@lamont-granquist that's a good point, I agree.

I have made a patch that adds attr_accessor to @node and @events of run_status so they can be set inside #build_node.

File: lib/chef/client.rb [hunk 2]
```diff
@@ -241,7 +241,8 @@ def load_node
     # Chef::Node:: The updated node object
     def build_node
       policy_builder.build_node
-      @run_status = Chef::RunStatus.new(node, events)
+      @run_status.node = node
+      @run_status.events = events
```
Comment (lamont-granquist) [inline lib/chef/client.rb]: the events object should be wired up correctly in the initializer already i think.

**Comment by lamont-granquist** [comment]
Yeah that looks better, i think that one line can probably be dropped, but that addresses my issue...

:+1:

**Comment by Igorshp** [comment]
It looked like @vents object was mutated at several places before it gets to #build_node.

I'm reading into how Eventdispatcher works to see if that line can be dropped.

**Comment by Igorshp** [comment]
@lamont-granquist 
Looks like subscribes in events get updated from here
https://github.com/chef/chef/blob/4827c9d31e68eafdff8d865d40479a18e6382793/lib/chef/client.rb#L434
that function will eventually lead to:
https://github.com/chef/chef/blob/4827c9d31e68eafdff8d865d40479a18e6382793/lib/chef/client.rb#L222

So it seems like that line is required.

**Comment by lamont-granquist** [comment]
Its the same events object though, so the reference that you're assigning should be the same.

**Comment by Igorshp** [comment]
@lamont-granquist 
I've done some tests to see what the value of @subscribers inside the @events is at various points and you are indeed right - the contents of @events inside main loop is same as inside @run_status. There is no need to set events again.

I'll update the PR shortly

**Comment by Igorshp** [comment]
@lamont-granquist do you have any other comments on this PR? how can we move this forward?

File: spec/unit/client_spec.rb [hunk 3]
```diff
@@ -819,4 +819,24 @@ def stub_for_audit
...
+
+    it "should run exception handlers on early fail" do
+      expect(subject).to receive(:run_failed)
+      exception = nil
+      begin
+        subject.run
+      rescue e => ex
+        exception = ex
+      end
+      expect(exception).to be_instance_of(e)
```
Comment (btm) [inline spec/unit/client_spec.rb]: Could this all be just ```expect { subject.run }.to raise_error(NoMethodError)```?

**Comment by Igorshp** [comment]
@btm I tried your suggestion, but I cannot get it to work. Could you show a full snippet of how it's supposed to work?
In the meantime, I have cleaned up the test a little to improve readability and reduce clutter.

**Comment by lamont-granquist** [comment]
Looks fine.  :+1:

**Comment by lamont-granquist** [comment]
Oh except, yeah you should use what @btm wrote.

Did you make sure to use curly braces in the `expect { }.to raise_error`?  Because that should do exactly what your code does...

**Comment by Igorshp** [comment]
@lamont-granquist bah, you're right, it was the curly braces! i was doing the standard expect().to
Updated the PR, thanks!

**Comment by btm** [comment]
:+1:

**Comment by Igorshp** [comment]
restarting appveyor/travis builds

**Pull request merged**